_posts/2019-09-10-unbearable-lightness-of-refactoring.markdown

```diff
@@ -0,0 +1,115 @@
+We found out that we can use `raise ActiveRecord::Rollback` which is a specific exception in Rails that only bubbles up in the surrounding transaction and does not get thrown outside of the block. This already makes things super complicated.
+
+In order to make our code less complicated, we did few things:
+
+* We delegated more responsibility to `OrderProcessor`
+* Instead of wrapping all code in one transaction, we now optimistically `submit` the order at the beginning and in case anything went wrong, we revert the changes.
+
+```ruby
+order_processor = OrderProcessor.new(order, user_id)
+raise Errors::ValidationError, order_processor.validation_error unless order_processor.valid?
+
+order_processor.advance_state(:submit!)
+unless order_processor.deduct_inventory
+  order_processor.revert!
+  raise Errors::InsufficientInventoryError
+end
+
+order_processor.set_totals!
+order_processor.hold
+order_processor.store_transaction
+
+if order_processor.failed_payment?
+  order_processor.revert!
+  raise Errors::FailedTransactionError.new(:charge_authorization_failed, order_processor.transaction)
+elsif order_processor.requires_action?
+  order_processor.revert!
+  Exchange.dogstatsd.increment '******'
+  raise Errors::PaymentRequiresActionError, order_processor.action_data
+end
+order_processor.on_success
+```
+Well, this at least is lot more readable.
+
+
+### Get The Change to Production
+
+Next question is how to get this to production. We tried to isolate this specific refactoring by:
+
+* Open a PR that only focuses on our refactoring
+* Make sure in the PR above we don't touch any API level tests and make sure all these tests still pass. This would give us more confidence that we are not impacting our existing clients.
+* Review and merge refactoring PR and test on staging.
+* Deploy everything in current pipeline to isolate the refactor deploy.
+* Deploy the refactoring PR to production.
+
+## How Did It Go?
+
+This plan worked for us, for the most part. We ended up having to rollback the deploy since we found a bug in a non-API part of our app. From this we learned that even if we already have tests written in different layers of our app, we still need to verify them and make sure they cover all cases. Relying too much on existing tests can often lead to trouble -- verify!
+
+## Our learnings
+- Don't be afraid of refactors. They are natural and a healthy engineering tool / practice.
+- Ensure that refactor PR's only include refactor-related changes. It's often tempting to fix other things along the way, but those fixes can take place in follow-up PRs.
+- Don't only rely on existing tests. Refactoring is great opportunity to review and verify your tests. Verify them and make sure they cover all scenarios.
+
+Curious about the PR? At Artsy we believe in Open Source By Default, so check out the code [here](https://github.com/artsy/exchange/pull/475/files).
```

Comment:
This is awesome! We should add a link to the Open Source by Default section of our principles doc: https://github.com/artsy/README/blob/master/culture/engineering-principles.md#open-source-by-default

Also, what do you think about adding a link to Exchange's repo near the beginning of the post? Just to make it easy for someone skimming to find?

Reply:
good point, where would you add the link to exchange, in some ways i didn't want to distract people with repo :)

Reply:
Somewhere near the top, since that's where people will probably go to look for it. You can add it after the quote, maybe, in a paragraph surrounded by parens to denote its unimportance to the narrative of the post:

```markdown
> (If you're looking for the code that this blog post refers to, [the Exchange repo is on GitHub](https://github.com/artsy/exchange).)
```